File: dsa/concepts/05_binarysearch/java/L33_SearchInRotatedSortedArray.java
public class L33_SearchInRotatedSortedArray {

	public static void main(String[] args) {
		L33_SearchInRotatedSortedArray obj = new L33_SearchInRotatedSortedArray();
		System.out.println(obj.search(new int[] {4,5,6,7,8,1,2,3}, 8));     // 4
	}

    // Problem: https://leetcode.com/problems/search-in-rotated-sorted-array/
    // Idea: Find sorted portion of array and check if target present in that range as you can check its availability only in sorted portion of array.
    // If yes, do bs on left portion, else do bs on right portion
	public int search(int[] nums, int target) {
        int low = 0, high = nums.length-1;
        while(low <= high){
            int mid = (low + high) / 2;
            if (nums[mid] == target)
                return mid;

            if (nums[low] <= nums[mid]) {   // check if left sorted
                if (nums[low] <= target && target <= nums[mid])
                    high = mid - 1;
                else
                    low = mid + 1;
            } else {                        // right sorted
                if (nums[mid] <= target && target <= nums[high])
                    low = mid + 1;
                else
                    high = mid - 1;
            }
        }
        return -1;
    }

}

File: dsa/concepts/05_binarysearch/python/l33_search_in_rotated_sorted_array.py
"""
LeetCode 33 - Search in Rotated Sorted Array

Idea:
At every step, one half of the array is always sorted.
Check which half is sorted and decide where the target can lie.

Time Complexity: O(log n)
Space Complexity: O(1)
"""

def search(nums, target):
    low, high = 0, len(nums) - 1

    while low <= high:
        mid = (low + high) // 2

        if nums[mid] == target:
            return mid

        # Left half is sorted
        if nums[low] <= nums[mid]:
            if nums[low] <= target < nums[mid]:
                high = mid - 1
            else:
                low = mid + 1
        # Right half is sorted
        else:
            if nums[mid] < target <= nums[high]:
                low = mid + 1
            else:
                high = mid - 1

    return -1


# Driver code
if __name__ == "__main__":
    nums = [4, 5, 6, 7, 8, 1, 2, 3]
    target = 8
    print(search(nums, target))   # 4
